Automattic/woocommerce-payments
# Limit amount of Apple Pay verification requests

Fixes #3045

### Changes proposed in this Pull Request

We've seen instances of spikes on Apple Pay domain verification requests. This PR aims to fix these spikes on two distinct scenarios:
- p1635413607201000-slack-CGGCLBN58 A spike right after release - probably due to too many merchants updating the plugin and triggering the verification request. This is due to the way the verification was performed right after update in case the domain association file had changed. We're trying to fix this here by only triggering a request if the file actually changed.
- p1633438592224000-slack-CGGCLBN58 We're not 100% sure why this is, but one merchant was able to trigger the domain verification request hundreds of times. This is perhaps due to opening the website via two separate domains where the `$_SERVER` superglobal would return whatever is the current domain URL, instead of the domain that's configured to be the canonical one. In this example, we had a merchant trigger the spike in requests probably because they had both `example.wordpress.com` and `example.com` domain names. The use of `get_site_url()` should fix that.

### Testing instructions

- For Apple Pay domain verification to work, WCPay server must be configured with a live restricted key for Apple Pay, and WCPay dev utils must have dev mode disabled. Feel free to use my personal live account for this `acct_...ZoSU` (WCPay Dev). Found here: paJDYF-18Y-p2#comment-5116.

#### Testing verification upon domain domain change

- Comment out these lines from `docker/wordpress/wp-config.php`, as they're only used to provide support for Jurassic tube and ngrok, and won't really apply for most merchants who have their site URL configured as an option:
```php
define( 'WP_SITEURL', DOCKER_REQUEST_URL );
define( 'WP_HOME', DOCKER_REQUEST_URL );
```
- Start a ngrok or JT session for a publicly accessible URL, and add that URL to `wp_options`:
```sql
INSERT INTO `wp_options` (`option_id`, `option_name`, `option_value`, `autoload`)
VALUES (1, 'siteurl', 'https://xyz.jurassic.tube', 'yes');
```
- (The above can also be achieved by setting up a URL during installation or via **wp-admin > Settings > Site Address (URL)**).
- Navigate to **Payments > Settings** and make sure Express checkouts are enabled.
- Open a product page in Safari and make sure the Apple Pay button is visible.
- Repeat the steps above, but with a different domain name, and make sure the Apple Pay button is visible.
- That means the domain verification worked.

#### Testing verification upon update

The test below will only provide proof that it worked if you change the contents of the `apple-developer-merchantid-domain-association` file and add a breakpoint on the `verify_domain_if_configured` function.

Note: If a domain name was successfully verified in the past, even if you try to verify it with a wrong domain association file, it might still remain verified.

Don't worry if that doesn't make sense. You'll probably understand the changes and the expected result with the testing instructions and by looking at the change on the `woocommerce_woocommerce_payments_updated` filter. We're basically trying to avoid unnecessary verification requests if the file hasn't changed.

- Run the code below to simulate that you're on an older version of WCPay:
```sql
UPDATE wp_options SET option_value='3.2.0' WHERE option_name='woocommerce_woocommerce_payments_version';
```
- Navigate to any `wp-admin` page. WCPay will act as if the plugin was just updated, and will trigger a verification request if the domain association file has changed.

-------------------
